File: src/store/state.ml
type albums = {
  baseById : (string, Subsonic.BaseAlbum.t) Js.Map.t;
  detailsById : (string, Subsonic.AlbumDetails.t) Js.Map.t;
  infoById : (string, Subsonic.AlbumInfo.albumInfo) Js.Map.t;
  songIdsById : (string, string array) Js.Map.t;
}

type artists = {
  albumIdsByArtistId : (string, string array) Js.Map.t;
  artistInfoById : (string, Subsonic.ArtistInfo.withoutSimilarArtist) Js.Map.t;
  byId : (string, Subsonic.BaseArtist.t) Js.Map.t;
  listIds : string array option;
  similarArtistsById : (string, Subsonic.SimilarArtist.t array) Js.Map.t;
  topSongIdsByArtistName : (string, string array) Js.Map.t;
}

type auth = {
  clearSubsonicCredentials : unit -> unit;
  credentials : Subsonic.Credentials.t option;
  saveSubsonicCredentials : Subsonic.Credentials.t -> unit;
}

module ReplayGainOptions = struct
  open Ppx_deriving_json_runtime.Primitives

  type[@deriving json] preferredGain = [ `album | `track ]

  let stringOfPreferredGain = function `album -> "album" | `track -> "track"

  let preferredGainOfString = function
    | "album" -> Some `album
    | "track" -> Some `track
    | _ -> None

  type[@deriving json] t = {
    preAmp : float;
    preferredGain : preferredGain option; [@json.option]
  }
end

type timeRange = { end_ : float; [@mel.as "end"] start : float }

type player = {
  buffered : timeRange Js.Array.t;
  currentSongId : string option;
  currentTime : float;
  duration : float option;
  muted : bool;
  paused : bool;
  queuedSongIds : string Js.Array.t;
  replayGainOptions : ReplayGainOptions.t;
  volume : float;
}

type songs = { byId : (string, Subsonic.Song.t) Js.Map.t }

type t = {
  albums : albums;
  artists : artists;
  auth : auth;
  player : player;
  songs : songs;
}

File: src/store/state.ts
import * as v from 'valibot';

import type {
  AlbumInfo,
  ArtistInfoWithoutSimilarArtist,
  BaseAlbum,
  BaseArtist,
  Credentials,
  SimilarArtist,
  Song,
} from '../subsonic';
import { DiscTitle, ReleaseDate } from '../subsonic';

const AlbumDetails = v.object({
  discTitles: v.optional(v.array(DiscTitle)),
  isCompilation: v.optional(v.boolean()),
  originalReleaseDate: v.optional(ReleaseDate),
  releaseDate: v.optional(ReleaseDate),
});

type AlbumDetails = v.InferOutput<typeof AlbumDetails>;

export const ReplayGainOptions = v.object({
  preAmp: v.number(),
  preferredGain: v.optional(v.union([v.literal('album'), v.literal('track')])),
});

export type ReplayGainOptions = v.InferOutput<typeof ReplayGainOptions>;

export interface StoreState {
  albums: {
    baseById: Map<string, BaseAlbum>;
    detailsById: Map<string, AlbumDetails>;
    infoById: Map<string, AlbumInfo>;
    songIdsById: Map<string, string[]>;
  };
  artists: {
    albumIdsByArtistId: Map<string, string[]>;
    artistInfoById: Map<string, ArtistInfoWithoutSimilarArtist>;
    byId: Map<string, BaseArtist>;
    listIds?: string[];
    similarArtistsById: Map<
      string,
      Array<
        | { TAG: 0; _0: Extract<SimilarArtist, { id: unknown }> }
        | { TAG: 1; _0: Exclude<SimilarArtist, { id: unknown }> }
      >
    >;
    topSongIdsByArtistName: Map<string, string[]>;
  };
  auth: {
    clearSubsonicCredentials: () => void;
    credentials?: Credentials;
    saveSubsonicCredentials: (credentials: Credentials) => void;
  };
  player: {
    buffered: Array<{ end: number; start: number }>;
    currentSongId?: string;
    currentTime: number;
    duration?: number;
    muted: boolean;
    paused: boolean;
    queuedSongIds: string[];
    replayGainOptions: ReplayGainOptions;
    volume: number;
  };
  songs: { byId: Map<string, Song> };
}
